### ISSUE TITLE
Allow username display to be case-insensitive

### ISSUE THREAD

**Issue by raisedadead** [freeCodeCamp/freeCodeCamp#35525]
- We want to store a case-sensitive version of usernames for display both in the URL and on the profile itself. 
   This is how GitHub, Twitter, and many other services operate.

Twitter example: 

<img width="461" alt="bri_bri_ _on_twitter___someone_sent_me_a_thank_you_note_for_helping_to_inspire_them__she_comes_to_the_javascript_study_group_every_thursday__this_just_made_my_day_ _https___t_co_aiuapfmtgf_" src="https://user-images.githubusercontent.com/985197/51850165-29fe3600-235c-11e9-89f6-4cf98fae422a.png">

What this means as a fix:

- [ ] Add an additional key `usernameDisplay` in the user schema:

   ```js
	{
	...
	"properties": {
	    ...
	    "username": {
	      "type": "string",
	      "index": {
	        "mongodb": {
	          "unique": true,
	          "background": true
	        }
	      },
	      "require": true
	    },
       "usernameDisplay": {
		  "type": "string",
		  ...
		}
    ...
	```

- [ ] This new `usernameDisplay` key will default to `user.username` on user creation.
- [ ] When a user updates their user name in the settings page, we fetch & update the `usernameDisplay` key. We also update the `username` key to its lowercase equivalent from the `usernameDisplay` key in the api-server side logic.
- [ ] The URL lookup and  for a username should still default to the `username` lowercase.
- [ ] When the profile page is rendered, the display will render the `usernameDisplay` instead, but should fallback to the username in case its `null` or empty (existing users will not have this key on db)

_Originally posted by @QuincyLarson in https://github.com/freeCodeCamp/freeCodeCamp/issues/19346#issuecomment-458199335_

**Comment by icatat** [comment]
Hi! I am interested in working on this issue!

**Comment by salma71** [comment]
I'm interested too. If you like we can work together

**Comment by seki2020** [comment]
> I'm interested too. If you like we can work together

Do I still have opportunity to join this issue?

**Comment by icatat** [comment]
@salma71 working together sounds good, let me know if there is anyway I could contact you

**Comment by DanieSegarra36** [comment]
Has someone started working on this issue?

**Comment by icatat** [comment]
@DanieSegarra36  @addv12345 and I are currently working on it

**Comment by piankot** [comment]
A bit confused but has this been implemented into the live website yet?

**Comment by ojeytonwilliams** [comment]
Good question, @piankot. It was briefly on the development server, but there were some problems so it was reverted in https://github.com/freeCodeCamp/freeCodeCamp/pull/37915.

We're still planning to implement it, but some more work is needed.

**Comment by MarceloDamian** [comment]
Hello! Would it be possible to work on this issue!?

**Comment by ojeytonwilliams** [comment]
Hey @MarceloDamian we ran into problems last time this was implemented, discussed here: https://github.com/freeCodeCamp/freeCodeCamp/pull/37915.

Before we try this again, I'd like to have e2e tests that can catch those problems.  I don't have an ETA for that, but I will be working on it this week.  After they are done, I think we can try this again.

**Comment by MarceloDamian** [comment]
Sounds good thank you!

**Comment by deepgoonumich** [comment]
Hello, my colleague @manrajtoor and I also wanna take a shot at this for a school project. Excited to work an learn from everyone in this community.

**Comment by ojeytonwilliams** [comment]
You can track the progress on the e2e testing setup here: https://github.com/freeCodeCamp/freeCodeCamp/pull/37053/

It's working well, but needs a few minor changes and to go through review.

**Comment by ojeytonwilliams** [comment]
@raisedadead Shaun pointed out that this is in conflict with https://github.com/freeCodeCamp/freeCodeCamp/issues/39697 is the behaviour described here what we want?

**Comment by raisedadead** [comment]
Just to clear out any confusion:

#39697 - This is an existing bug that doesn't tell the users we are defaulting to lowercase in the DB. That is they are think we are using the case that they enter, but we aren't.

#35525 - This is a feature request - which will supersede the bug above if implemented, that is, it allows for us to store an additional property which stores the username in format that the user wants. The reason we want to store that separately and not on the current `username` property is to avoid breaking anything anywhere else.

Simply put, `username` will have a sole responsibility of being used as a key and the addition property that this issue describes will have a display only variant of the `username` that user prefers to see on their portfolio.

**Comment by ojeytonwilliams** [comment]
Thanks, that makes sense.  So, both issues are still valid and open to be worked on.

**Comment by veedata** [comment]
Hello, could I take up this issue?

**Comment by ojeytonwilliams** [comment]
Hi @veedata, yes, definitely.  We have some [contribution docs](https://contribute.freecodecamp.org/#/) that help with getting started.
